File: C/003-largest-prime-factor.c

/* 
    The prime factors of 13195 are 5, 7, 13 and 29.
    The largest among them is 29
    Write a program to find the largest prime factor of a number?
*/

#include<stdio.h>
#include<math.h>

int prime(int number){  //For letting in only prime numbers inside loop
    for (int i=3; i<=number/2; ++i){
        if(!(number % i == 0)){
            return 1;
        }
        else{
            return 0;
            break;
        }
    }
}

int main(){
    long limit, maxPrimeFactor, i;
    scanf("%ld", &limit);
    /*
    First divide the number by 2 for highest possible times in order to reduce it.
    */
    while(limit % 2 == 0){
        limit /= 2;
        maxPrimeFactor = 2;
    }
    /* 
    It is a convention that the prime factor would always be less than or equal to square root of the number 
    */
    for(i=3; i <= sqrt(limit); i = i+2){  //Increment by 2 for getting only odd numbers
        if(prime(i)){
            while(limit % i == 0){
                maxPrimeFactor = i;
                limit /= i;
            }
        }
    }
    if(limit > 2 && prime(limit)){
        maxPrimeFactor = limit;
    }
    printf("%ld", maxPrimeFactor);
    return 0;
}

File: C#/003-largest-prime-factor.cs
using System;

namespace LargestPrimeFactor
{
    class Program {

        public static void Main(string[] args)
        {
            Console.WriteLine(MaxPrimeFactor(600851475143));
        }

        // return the largest prime factor
        static private long MaxPrimeFactor (long n)
        {
            // n = x * y
            // so we start x = 1; then y is the largest facor
            // if y is also a prime that is the ans (largest prime factor)
            // this is the fastest way to get the ans
            for (long k = 1; k <= n/2; k++)
            {
                if (n % k == 0 && IsPrime(n/k)) // if k is a factor of n, then n/k is also a facor; and n/k is larger than k
                {
                    return n/k;
                }
            }
            return n;
        }

        // checks whether this number is a prime or not
        static private bool IsPrime(long x)
        {
            for (long i = 2; i < Math.Sqrt(x); i++)
            {
                if (x % i == 0) // if there are any divisibles other than 1 and itself this is not a prime
                {
                    return false;
                }
            }
            return true;
        }
    }
}
